addons/account_edi/models/account_edi_format.py

```diff
@@ -298,25 +288,18 @@ def _embed_edis_to_pdf(self, pdf_content, invoice):
         :param invoice: the invoice to generate the EDI from.
         :returns: the same pdf_content with the EDI of the invoice embed in it.
         """
-        attachments = []
-        for edi_format in self.filtered(lambda edi_format: edi_format._is_embedding_to_invoice_pdf_needed()):
-            attach = edi_format._get_embedding_to_invoice_pdf_values(invoice)
-            if attach:
-                attachments.append(attach)
-
-        if attachments:
-            # Add the attachments to the pdf file
-            reader_buffer = io.BytesIO(pdf_content)
-            reader = OdooPdfFileReader(reader_buffer, strict=False)
-            writer = OdooPdfFileWriter()
-            writer.cloneReaderDocumentRoot(reader)
-            for vals in attachments:
-                writer.addAttachment(vals['name'], vals['datas'])
-            buffer = io.BytesIO()
-            writer.write(buffer)
-            pdf_content = buffer.getvalue()
-            reader_buffer.close()
-            buffer.close()
+        # Add the attachments to the pdf file
+        reader_buffer = io.BytesIO(pdf_content)
+        reader = OdooPdfFileReader(reader_buffer, strict=False)
+        writer = OdooPdfFileWriter()
+        writer.cloneReaderDocumentRoot(reader)
+        for edi_format in self:
+            edi_format._prepare_invoice_report(writer, invoice)
+        buffer = io.BytesIO()
+        writer.write(buffer)
+        pdf_content = buffer.getvalue()
+        reader_buffer.close()
+        buffer.close()
```

Comment:
This part is not very efficient because you are performing a read + write even there is nothing to embed inside the writer.
Also, `_get_edi_attachment` is iterating among all edi_documents. Imo, `_prepare_invoice_report` should take either the edi.document, either the ir.attachment as parameter.